crates/sncast/src/helpers/scarb_utils.rs

```diff
@@ -191,43 +192,69 @@ pub fn get_scarb_metadata_with_deps(
     execute_scarb_metadata_command(&command)
 }
 
-#[must_use]
-pub fn verify_or_determine_scarb_manifest_path(
+pub fn assert_manifest_path_exists(
     path_to_scarb_toml: &Option<Utf8PathBuf>,
-) -> Option<Utf8PathBuf> {
+) -> Result<Utf8PathBuf> {
     if let Some(path) = path_to_scarb_toml {
         assert!(path.exists(), "Failed to locate file at path = {path}");
     }
 
     let manifest_path = path_to_scarb_toml.clone().unwrap_or_else(|| {
-        get_scarb_manifest()
-            .context("Failed to obtain manifest path from scarb")
-            .unwrap()
+        get_scarb_manifest().expect("Failed to obtain manifest path from scarb")
     });
 
     if !manifest_path.exists() {
-        return None;
+        return Err(anyhow!(
+            "Path to Scarb.toml manifest does not exist = {manifest_path}"
+        ));
     }
 
-    Some(manifest_path)
+    Ok(manifest_path)
 }
 
-pub fn get_package_metadata<'a>(
-    metadata: &'a scarb_metadata::Metadata,
-    manifest_path: &'a Utf8PathBuf,
-) -> Result<&'a scarb_metadata::PackageMetadata> {
-    let manifest_path = canonicalize(manifest_path.clone())
-        .unwrap_or_else(|err| panic!("Failed to canonicalize {manifest_path}, error: {err:?}"));
-
-    let package = metadata
+fn get_package_metadata_by_name<'a>(
+    metadata: &'a Metadata,
+    package_name: &str,
+) -> Result<&'a PackageMetadata> {
+    metadata
         .packages
         .iter()
-        .find(|package| package.manifest_path == manifest_path)
+        .find(|package| package.name == package_name)
         .ok_or(anyhow!(
-            "Path = {} not found in scarb metadata",
-            manifest_path.display()
-        ))?;
-    Ok(package)
+            "Package {} not found in scarb metadata",
+            &package_name
+        ))
+}
+
+fn get_default_package_metadata(metadata: &Metadata) -> Result<&PackageMetadata> {
+    match metadata.packages.iter().collect::<Vec<_>>().as_slice() {
+        [package] => Ok(package),
+        [] => Err(anyhow!("No package found in metadata")),
+        _ => Err(anyhow!(
+            "More than one package found in metadata - specify package using --package flag"
+        )),
+    }
+}
+
+pub fn get_package_metadata(
+    manifest_path: &Utf8PathBuf,
+    package_name: &Option<String>,
+) -> Result<PackageMetadata> {
+    let metadata = get_scarb_metadata(manifest_path)?;
+    match &package_name {
+        Some(package_name) => Ok(get_package_metadata_by_name(&metadata, package_name)?.clone()),
+        None => Ok(get_default_package_metadata(&metadata)?.clone()),
+    }
+}
+
+pub fn get_main_package_metadata(manifest_path: &Utf8PathBuf) -> Result<PackageMetadata> {
+    let metadata = get_scarb_metadata(manifest_path)?;
+    let main_package = metadata
+        .packages
+        .into_iter()
+        .find(|package| package.manifest_path == metadata.workspace.manifest_path)
+        .expect("Failed to find main package in metadata");
```

Comment:
This will fail for virtual workspaces I believe
https://docs.swmansion.com/scarb/docs/reference/workspaces.html#virtual-workspace

Reply:
yup, you are right. I have rebased https://github.com/foundry-rs/starknet-foundry/pull/1572 on friday but did not checked its status, and its failing. trying to figure out a fix now.